metrics/reporter.go

```diff
@@ -0,0 +1,35 @@
+// Copyright 2017 Mark Nevill. All Rights Reserved.
+// See LICENSE for licensing terms.
+
+package http_metrics
+
+import (
+	"net/http"
+	"time"
+)
+
+// Called when a new request is to be tracked.
+type Reporter interface {
+	// Start tracking a new request
+	Track(req *http.Request) Tracker
+}
+
+// Receives events about a tracked request.
+type Tracker interface {
+	// The exchange has started. This is called immediately after Reporter.Track.
+	// On the client, this is called before any data is sent.
+	// On the server, this is called after headers have been parsed.
+	RequestStarted()
+	// The request body has been read to EOF or closed, whichever comes first.
+	// On the client, this is called when the transport completes sending the request.
+	// On the server, this is called when the handler completes reading the request, and may be omitted.
```

Comment:
`and may be omitted.` sounds bad.. Does it mean `RequestRead` can be omitted totally ? (no metrics produced because of that?)

Reply:
I can see why it can be omitted, someone can just not read the body fully... not sure if that is the problem we should be worried about

Reply:
I think this is fine, we can even track the clients which are not closing/reading fully the bodies (: